File: 147.insertion-sort-list.py
#
# @lc app=leetcode id=147 lang=python3
#
# [147] Insertion Sort List
#

# Time:  O(n^2)
# Space: O(1)


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# @lc code=start
class Solution:
    def insertionSortList(self, head: ListNode) -> ListNode:
        dummy = ListNode(-1)
        while head:
            next = head.next
            curr = dummy
            while curr.next and curr.next.val <= head.val:
                curr = curr.next
            head.next = curr.next
            curr.next = head
            head = next
        return dummy.next


# @lc code=end

File: 147.insertion-sort-list.js
/*
 * @lc app=leetcode id=147 lang=javascript
 *
 * [147] Insertion Sort List
 */

function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

// @lc code=start
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var insertionSortList = function (head) {
  const dummy = new ListNode(-1);
  while (head) {
    const next = head.next;
    let curr = dummy;
    while (curr.next && curr.next.val <= head.val) {
      curr = curr.next;
    }
    head.next = curr.next;
    curr.next = head;
    head = next;
  }
  return dummy.next;
};
// @lc code=end
